### ISSUE TITLE
Completion suggestions when import map is configured

### ISSUE THREAD

**Issue by cleytoncarvalho** [denoland/vscode_deno#261]
Autocomplete for absolute imports configured in import_map.json don't work.

My import_map.json:

```json
{
  "imports": {
    "~/": "./src/",
    "testing/": "https://deno.land/std@0.79.0/testing/"
  }
}
```

My .vscode/settings.json:

```json
{
  "deno.enable": true,
  "deno.unstable": true,
  "deno.lint": true,
  "[typescript]": {
    "editor.defaultFormatter": "denoland.vscode-deno",
  },
  "[typescriptreact]": {
    "editor.defaultFormatter": "denoland.vscode-deno",
  },
  "deno.import_intellisense_origins": {
    "https://deno.land": true
  },
  "deno.import_map": "./import_map.json"
}
```

**Screenshots**

![Screenshot from 2020-11-23 20-49-58](https://user-images.githubusercontent.com/7952866/100028390-91a53100-2dcd-11eb-959f-12340479dfbe.png)

**Versions**

vscode: 1.51.1
deno: 1.5.4
extension: 2.3.3

**Comment by kitsonk** [comment]
We will be adding back autocomplete imports soon to the extension.

**Comment by jsejcksn** [comment]
> soon

@kitsonk Is there a new roadmap? (Sorry if I missed it, but I don't see one.)

I'm particularly curious to see a summary/overview feature diff of pre- vs post- 3.x for this extension—it's not currently clear which things were removed intentionally vs didn't make it in yet, but are planned. Again, if I missed some announcement or similar, please just link me. Thanks!

**Comment by kitsonk** [comment]
Ok, we added back import completions, which is good, but "local" import completions do not work with the import map.  While the LSP recognises the modules and imports them correctly, it does not offer completion suggestions.

So this issue will be the enhancement of that.

**Comment by Oloompa** [comment]
It neither work for relative imports in same project (without import map)

**Comment by bartlomieju** [comment]
@crowlKats could you take on this issue?

**Comment by crowlKats** [comment]
@bartlomieju will do

**Comment by harrysolovay** [comment]
Just wanted to (respectfully) ping this issue. Import-map-friendly completion would make for some extraordinary DXs.

**Comment by bartlomieju** [comment]
It's in @crowlKats' backlog. We have a long tail of issues to fix in LSP, so I can't guarantee when this one will be fixed. @harrysolovay if you want to tackle this problem yourself, we can provide some pointers to get started.

**Comment by harrysolovay** [comment]
No promises, but I'd take a look. Some pointers would be greatly appreciated :)

**Comment by crowlKats** [comment]
I couldnt reproduce this, as it actually is working; only problem is the example provided in the original issue should say `deno.importMap` instead of `deno.import_map`.

**Comment by bartlomieju** [comment]
> I couldnt reproduce this, as it actually is working; only problem is the example provided in the original issue should say `deno.importMap` instead of `deno.import_map`.

Just to be clear - we both tried to reproduce the problem but are unable too. As long as `deno.importMap` is configured in `.vscode/settings.json` autocompletion works as expected.

**Comment by harrysolovay** [comment]
[Here is a reproduction of the problem](https://github.com/harrysolovay/lack_of_import_mapped_completion_repro). As described in the Readme:

- [The `capi.dev` service](https://github.com/paritytech/capi/tree/main/server) implements intellisense in accordance with the [Deno LS intellisense API](https://deno.land/manual@v1.29.1/advanced/language_server/imports).
- The repro's `import_map.json` redirects `capi/`-prefixed imports to their `https://capi.dev/@v0.1.0-beta.22/`-prefixed counterparts.

```json
{
  "imports": {
    "capi/": "https://capi.dev/@v0.1.0-beta.22/"
  }
}
```

- We get import completions from `https://capi.dev/@v0.1.0-beta.22/`-prefixed imports

<img width="1209" alt="Screen Shot 2022-12-26 at 3 05 15 PM" src="https://user-images.githubusercontent.com/4893548/209587678-d0f1b7ef-b725-4f8a-879f-9bf7077aba5d.png">

- We get no import completions from `capi/`-prefixed imports

<img width="1158" alt="Screen Shot 2022-12-26 at 3 05 31 PM" src="https://user-images.githubusercontent.com/4893548/209587684-70a930bd-ceec-499d-99f6-0260c1771306.png">

**Comment by bartlomieju** [comment]
Thanks @harrysolovay, we'll investigate it.

**Comment by crowlKats** [comment]
@harrysolovay I have not been able to reproduce it with your example. Here a showcase:

https://user-images.githubusercontent.com/13135287/209694238-4107655c-1dd0-4341-af2a-cddb7d230b91.mov

There seems to be a behaviour change between direct and importmap, however it does work

**Comment by harrysolovay** [comment]
That is indeed strange: the `proxy/` completion *should* span several other URLs (not just `wss:rpc.polkadot.io`). Perhaps those completions are grabbed from a local cache that wasn't cleared?

Also, why are the full URLs listed out, instead of the next-up directory name? These behavioral inconsistencies are shown in the following screen capture.

https://user-images.githubusercontent.com/4893548/209719854-244012c7-bdd1-49e4-a633-a8858630cfc6.mov

**Comment by tjjfvi** [comment]
Yeah, I believe the import map completion bases it off of the files that have been loaded into the project (i.e. from a different/previous import), like (I believe) it does when a server does not support import completions. Ideally, it should use the import completions provided by the server.

This would also be useful in cases where people e.g. import map `std/` to `https://deno.land/std/`, as it would provide completions for all `std/` modules, rather than just those that have already been imported.

**Comment by tjjfvi** [comment]
> if you want to tackle this problem yourself, we can provide some pointers to get started.

@bartlomieju I might take a look at this soon; [pointers](https://xkcd.com/138/) would be very much appreciated :)

**Comment by lensbart** [comment]
Not sure if this is the same issue, as it pertains to the auto-inserted import statement rather than the autocompleted location in the import clause — but I’d expect the auto-inserted import statement to import from the alias defined in the import map:

```jsonc
// .vscode/settings.json
{ "deno.importMap": "./import_map.json" }
```

```jsonc
// import_map.json
{ "imports": { "react": "https://esm.sh/react@18.2.0" } }
```

https://github.com/denoland/vscode_deno/assets/20876627/abbf5e86-d720-4f8d-8506-8b57c5efbb6d

**Comment by bombillazo** [comment]
Any update on this issue?

**Comment by bartlomieju** [comment]
@dsherret is this solved with your latest update to import map remapping?

**Comment by bartlomieju** [comment]
I believe this issue has been solved in Deno v1.35. Please let me know if you still hit a problem.
